cli/ballerina-packerina/src/main/java/org/ballerinalang/packerina/cmd/BuildCommand.java

```diff
@@ -77,26 +80,79 @@ public void execute() {
+        Path currentDir = Paths.get(System.getProperty(USER_DIR));
         if (argList == null || argList.size() == 0) {
             // ballerina build
-            BuilderUtils.compileWithTestsAndWrite(sourceRootPath, offline, lockEnabled, skiptests);
+            BuilderUtils.compileWithTestsAndWrite(currentDir, offline, lockEnabled, skiptests);
         } else {
             // ballerina build pkgName [-o outputFileName]
             String targetFileName;
-            Path sourcePath = Paths.get(argList.get(0));
+            String pkgName = argList.get(0);
+            if (pkgName.endsWith("/")) {
+                pkgName = pkgName.substring(0, pkgName.length() - 1);
+            }
             if (outputFileName != null && !outputFileName.isEmpty()) {
                 targetFileName = outputFileName;
             } else {
-                targetFileName = sourcePath.toString();
+                targetFileName = pkgName;
             }
 
-            BuilderUtils.compileWithTestsAndWrite(sourceRootPath, sourcePath.toString(), targetFileName,
-                                                  buildCompiledPkg, offline, lockEnabled, skiptests);
+            Path sourceFilePath = Paths.get(pkgName);
+            Path resolvedFullPath = currentDir.resolve(sourceFilePath);
+            // If the source is a single bal file which is not inside a project
+            if (Files.isRegularFile(resolvedFullPath) &&
+                    sourceFilePath.toString().endsWith(BLangConstants.BLANG_SRC_FILE_SUFFIX) &&
+                    !RepoUtils.hasProjectRepo(currentDir)) {
+                // If there is no output file name provided, then the executable (balx) should be created in the user's
+                // current directory with the same source file name with a balx extension. So if there's no output file
+                // name provided (with flag -o) and the target file name contains the full path (along with its parent
+                // path) and not only the source file name, then we extract the name of the source file to be the target
+                // file name
+                targetFileName = getTargetFileName(Paths.get(targetFileName));
+
+                // The source root path should be the parent of the source file
+                Path parent = resolvedFullPath.getParent();
+                currentDir = parent != null ? parent : currentDir;
+
+                // The package name/source should be the source file name
+                Path resolvedFileName = resolvedFullPath.getFileName();
+                pkgName = resolvedFileName != null ? resolvedFileName.toString() : pkgName;
+
+            } else if (Files.isDirectory(currentDir)) { // If the source is a package from a project
+                // Checks if the source is a package and if its inside a project (with a .ballerina folder)
+                if (Files.isDirectory(resolvedFullPath) && !RepoUtils.hasProjectRepo(currentDir)) {
```

Comment:
This if condition can merged with the parent else if condition

Reply:
Merged with d674cdf